pkg/util/workqueue/telemetry/telemetry.go

```diff
@@ -0,0 +1,229 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+// Package telemetry provides implementation for the workqueue metrics provider
+package telemetry
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/DataDog/datadog-agent/pkg/telemetry"
+)
+
+var (
+	commonOpts = telemetry.Options{NoDoubleUnderscoreSep: true}
+)
+
+type gaugeWrapper struct {
+	telemetry.Gauge
+}
+
+// Inc implements the workqueue.GaugeMetric interface
+func (g gaugeWrapper) Inc() {
+	g.Gauge.Inc()
+}
+
+// Dec implements the workqueue.GaugeMetric interface
+func (g gaugeWrapper) Dec() {
+	g.Gauge.Dec()
+}
+
+// Set implements the workqueue.GaugeMetric interface
+func (g gaugeWrapper) Set(v float64) {
+	g.Gauge.Set(v)
+}
+
+type counterWrapper struct {
+	telemetry.Counter
+}
+
+// Inc implements the workqueue.CounterMetric interface
+func (c counterWrapper) Inc() {
+	c.Counter.Inc()
+}
+
+type histgramWrapper struct {
+	telemetry.Histogram
+}
+
+// Observer implements the workqueue.HistogramMetric interface
+func (l histgramWrapper) Observe(value float64) {
+	l.Histogram.Observe(value)
+}
+
+// QueueMetricsProvider is a workqueue.MetricsProvider that provides metrics for the SBOM queue
+type QueueMetricsProvider struct {
+	mutex   *sync.Mutex
+	metrics map[string]interface{}
+}
+
+// NewQueueMetricsProvider returns a new Queue Metrics Provider
+func NewQueueMetricsProvider() QueueMetricsProvider {
+	return QueueMetricsProvider{
+		mutex:   &sync.Mutex{},
+		metrics: map[string]interface{}{},
+	}
+}
+
+// Ensure QueueMetricsProvider implements the workqueue.MetricsProvider interface
+var _ workqueue.MetricsProvider = QueueMetricsProvider{}
+
+type initializer[T any] func(subsystem, name, description string) T
+
+func register[T any](q QueueMetricsProvider, metricName, subsystem, description string, init initializer[T]) T {
+	if q.mutex == nil {
+		panic("Queue Metrics Provider mutex lock should not be nil")
+	}
+
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	if metric, found := q.metrics[metricName]; found {
+		return metric.(T)
```

Comment:
Perhaps we should have a log here to prevent misusage. What do you think ?

Reply:
Are you referring to type assertion?

Reply:
I'm referring about the fact that we don't create a new metric because it already exists.